### PR TITLE
Redoes https://github.com/eclipse/deeplearning4j/pull/9514

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?
Fix several tests including conditions, scatter add diff, softmax coss entropy loss
Focuses on test fixes relating to gradient checks.
Mostly java serialization related failures.
When we train a model, we create a gradient function which copies the graph
and appends a set of gradient ops for execution for our autodiff.
This graph created uses our flatbuffers support to copy graph configurations
from the base graph (the feed forward) to the training graph.

When we serialize models, various configurations and java op classes need to bootstrap configuration specific objects such as Conv2DConfig. This introduces several fixes related to test failures that were caused by serialization problems.

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [X ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X ] Created tests for any significant new code additions.
- [X ] Relevant tests for your changes are passing.

### CODE REVIEW

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/Choose.java [hunk 1]
```diff
@@ -45,7 +45,7 @@ public Choose(SameDiff sameDiff, SDVariable[] args, Condition condition) {
 
         this.inPlace = true;
         this.inplaceCall = true;
-        addIArgument(condition.condtionNum());
+        addIArgument(condition.conditionType().ordinal());
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/Choose.java]: still using ordinal here.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/Choose.java [hunk 2]
```diff
@@ -60,7 +60,7 @@ public Choose(String opName, INDArray[] inputs, Condition condition) {
         }
 
         addInputArgument(inputs);
-        addIArgument(condition.condtionNum());
+        addIArgument(condition.conditionType().ordinal());
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/Choose.java]: still using ordinal here.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/Choose.java [hunk 3]
```diff
@@ -96,7 +96,7 @@ public Choose(INDArray[] inputs,List<Integer> iArgs, List<Double> tArgs,Conditio
 
         if(!tArgs.isEmpty())
             addTArgument(Doubles.toArray(tArgs));
-        addIArgument(condition.condtionNum());
+        addIArgument(condition.conditionType().ordinal());
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/Choose.java]: still using ordinal here.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/indexing/conditions/Conditions.java [hunk 4]
```diff
@@ -24,6 +24,59 @@
...
+                case 7: return ABS_GREATER_THAN;
+                case 8: return IS_INFINITE;
+                case 9: return IS_NAN;
+                case 10: return ABS_EQUALS;
+                case 11: return NOT_EQUALS;
+                case 12: return ABS_GREATER_OR_EQUAL;
+                case 13: return ABS_LESS_THAN_OR_EQUAL;
+                case 14: return IS_FINITE;
+                case 15: return NOT_FINITE;
+                case 16: return AGGREGATE;
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/indexing/conditions/Conditions.java]: should be -1.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/indexing/conditions/Conditions.java [hunk 5]
```diff
@@ -24,6 +24,59 @@
...
+        ABS_GREATER_THAN(7), //7
+        IS_INFINITE(8), //8
+        IS_NAN(9), //9
+        ABS_EQUALS(10), //10
+        NOT_EQUALS(11), //11
+        ABS_GREATER_OR_EQUAL(12), //12
+        ABS_LESS_THAN_OR_EQUAL(13), //13
+        IS_FINITE(14), //14
+        NOT_FINITE(15), //15
+        AGGREGATE(-1); // this is an aggregate enum for or, and, not, and other indirect conditions that depend on others
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/indexing/conditions/Conditions.java]: No need to have the value commented after it anymore.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/indexing/conditions/Conditions.java [hunk 6]
```diff
@@ -305,4 +382,54 @@ public static Condition absLessThan(Number value) {
...
+
+    public static Condition fromInt(int mode) {
+        return fromInt(mode,0.0);
+    }
+
+
+    public static Condition fromInt(int mode,Double value) {
+        if(value == null) value = 0.0;
+
+        switch(ConditionMode.values()[mode]) {
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/indexing/conditions/Conditions.java]: should use `ConditionMode.fromNumber`

**Review by treo — CHANGES_REQUESTED** [review]
A few more subtle bugs

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/indexing/conditions/Conditions.java (hunk 5 above)
Comment (agibsonccc) [reply nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/indexing/conditions/Conditions.java]: I actually disagree here and will keep it. The "aggregate" is a catch all for anything that's not a direct evaluation.

Comment (treo) [reply nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/indexing/conditions/Conditions.java]: I'm taking about all the things above it.

**Review by treo — APPROVED** [review]

**Pull request merged**